### DISCUSSION TITLE
Vite CSP support

### DISCUSSION THREAD

**Discussion by dargmuesli** [Baroshem/nuxt-security#454] [Ideas]
Hey @vejja,

I think we can add a meta header that resolves issues with Vite and CSP now:

- https://github.com/vitejs/vite/pull/16052 is merged now
  - Vite now checks for a meta header with `property="csp-nonce" nonce="..."` ([docs](https://vitejs.dev/guide/features.html#content-security-policy-csp))
- my proposal is to add something similar to this
  ```ts
  export default defineNitroPlugin((nitroApp) => {
    if (import.meta.dev) {
      nitroApp.hooks.hook('render:html', (html, { event }) => {
        html.head.push(
          `<meta property="csp-nonce" nonce="${event.context.security.nonce}">`,
        )
      })
    }
  }
  ```
  around https://github.com/Baroshem/nuxt-security/blob/8305b48079dc7c0ef83a16d3f6de24dfc41e563b/src/runtime/nitro/plugins/80-cspSsrNonce.ts#L23
  maybe in a style similar to this: https://github.com/Baroshem/nuxt-security/blob/8305b48079dc7c0ef83a16d3f6de24dfc41e563b/src/runtime/nitro/plugins/60-cspSsgHashes.ts#L108
  I haven't used cheerio before, therefore I'm not sure.
- this could make this warning superfluous:
  https://github.com/Baroshem/nuxt-security/blob/8305b48079dc7c0ef83a16d3f6de24dfc41e563b/docs/content/1.documentation/2.headers/1.csp.md?plain=1#L222

## What do you think? :tada:

cc @Baroshem of course

**Comment by Baroshem** [comment]
Hey @dargmuesli 

Thanks for this interesting idea! I like where this js going but I would wait for @vejja who is our expert in this area to share his feedback :)

**Reply by vejja** [reply]
Hi @dargmuesli 
I am not knowledgeable about Vite at all to be honest.
Could you give a bit more details on what is the purpose of this Vite feature and how it works ?
- do we need to set the `html.cspNonce` value somewhere ?
- is it only for dev mode ?
- are we supposed to write the `meta` tag or read it ?
- will Vite also control the nonce in the headers ?
- do we still need the `nonce-{{nonce }}` placeholders ?
- how do we modify the nonce value in the CSP directives then ?
- how can this help solve the `style-src: unsafe-inline` issue ? 

I'm trying to understand how this plays with the Nuxt build, but overall sounds awesome if we can offload this to a native solution.

**Reply by dargmuesli** [reply]
No problem, sure!

> do we need to set the `html.cspNonce` value somewhere ?

I think in the Nuxt build process Vite's `html.cspNonce` setting is not used by the Vue compilation tools so changing it on our end would currently have no effect.

> is it only for dev mode ?

I think so, yes.

> are we supposed to write the `meta` tag or read it ?

Write it, e.g. as proposed in my post above. Vite will then respect this on the client side, resulting in no CSP errors in the browser.

> will Vite also control the nonce in the headers ?

Vite will not interact with http headers, as to my knowledge.

> do we still need the `nonce-{{nonce }}` placeholders ?

Yes. The changes to Vite only make Vite respect the currently defined CSP in our case.

> how do we modify the nonce value in the CSP directives then ?

No need to do so due to the same reason.

> how can this help solve the `style-src: unsafe-inline` issue ?

As I understand it, this directive was necessary for development only as Vite modifies styles client side and has previously not been doing so without breaking CSP. Now it respects CSP by reading the `meta` so I think `style-src: unsafe-inline` is no longer needed. If there are other reason to have `style-src: unsafe-inline` then it should stay of course.

**Reply by Baroshem** [reply]
That sounds really interesting @dargmuesli !

Do you think you could create a Proof of Concept PR of how this would work?

I would love to see how we could utilize native vite solution for that :)

**Reply by dargmuesli** [reply]
Right now I'm focused on my master's thesis and the event I'm in contact with you about via Discord. Do you think the `cheerios.head.unshift` method I mentioned in my original post would be the correct way to add the `<meta property="csp-nonce" nonce="${event.context.security.nonce}">`?

**Reply by Baroshem** [reply]
Well, we recently removed the cheerios in favor of regular regexp so I am not sure. Maybe @GalacticHypernova or @vejja would be able to answer it better.

And no worries about that. We can come back to this topic after the event and thesis. Thanks for letting us know. I also do have some time consuming activities in the schedule so completely understand :D

**Reply by GalacticHypernova** [reply]
What is the priority of this meta tag, if it has a priority? If it just needs to be somewhere in the head we can easily just push it, regardless of cheerio

**Reply by dargmuesli** [reply]
Afaik it can go anywhere in `<head>`.

**Reply by GalacticHypernova** [reply]
Well then that's very easy to implement!
A simple 
```ts
html.head.push(`<meta property="csp-nonce" nonce="{the nonce}">`)
```

**Reply by dargmuesli** [reply]
Created PR https://github.com/Baroshem/nuxt-security/pull/475
